src/Vibrato/Vibrato.cpp

```diff
@@ -0,0 +1,135 @@
+#include "Vibrato.h"
+
+VibratoEffector::VibratoEffector(int iSampleRate, int iNumChannel, float fModulationFreqInHz, float fModulationWidthInS)
+    : m_iSampleRate(iSampleRate), 
+      m_iNumChannel(iNumChannel),
+      m_fModulationFreqInHz(fModulationFreqInHz), 
+      m_fModulationWidthInS(fModulationWidthInS) {
+    updateDelayBuffer();
+    updateLfo();
+}
+
+void VibratoEffector::updateDelayBuffer() {
+    deleteBuffer();
+    m_fDelayInSample = m_fModulationWidthInS * static_cast<float>(m_iSampleRate);
+    int iDelayInSample = static_cast<int>(ceil(m_fDelayInSample * 3 + 2));
+
+    int DELAY = static_cast<int>(round(m_fModulationWidthInS * static_cast<float>(m_iSampleRate)));
+    int WIDTH = DELAY;
+    int L = 2 + DELAY + WIDTH * 2;
+
+    m_buffer = new CRingBuffer<float>*[m_iNumChannel];
+    for (int i = 0; i < m_iNumChannel; ++i) {
+        m_buffer[i] = new CRingBuffer<float>(L);
+        m_buffer[i]->setReadIdx(0);
+        m_buffer[i]->setWriteIdx(L);
+    }
+}
+
+void VibratoEffector::deleteBuffer() {
+    if (m_buffer != nullptr) {
+        for (int i = 0; i < m_iNumChannel; ++i) {
+            if (m_buffer[i] != nullptr) {
+                delete m_buffer[i];
+            }
+        }
+        delete[] m_buffer;
+        m_buffer = nullptr;
+    }
+}
+
+void VibratoEffector::updateLfo() {
+    deleteLfo();
+    m_lfo = new CLfo*[m_iNumChannel];
+    for (int i = 0; i < m_iNumChannel; ++i) {
+        m_lfo[i] = new CLfo(m_iSampleRate, 1 << 12, 1, m_fModulationFreqInHz); // TODO: how to determine the table size?
```

Comment:
is a new wavetable created every time the LFO frequency changes?

Reply:
No. `updateLfo()` is only called in the `VibratoEffector` constructor. When LFO frequency changes, `setParam` will dispatch parameter into [`setModulationFreq`](https://github.com/medioqrity/2022-MUSI6106/blob/assignment2_vibrato/src/Vibrato/Vibrato.cpp#L98), which does not call `updateLfo()`.